### DISCUSSION TITLE
osu! Mod Proposal: Strict Tracking

### DISCUSSION THREAD

**Discussion by apollo-dw** [ppy/osu#17130] [Ideas]
This is a proposal for a difficulty increasing mod called Strict Tracking (or some other alternative). This mod would move slider tracking judgement to be moved away from slider ticks, and to keeping the slider follow circle active throughout the entire duration of the slider. Misaiming or letting go off of the sliderball would therefore result in a combo break. This would increase the difficulty of maps with slider elements in, especially those with complex sliderbodies and fast slider velocities that are usually cheesed by just aiming the sliderticks.

I'd be happy to work towards implementation - I would just need some input towards the "how" side of things. At the moment, the plan would be to:

- turn off slider tick generation / remove nested ticks in Slider objects when the mod is active,
- in an `Update(Playfield playfield)` foreach aliveobjects block, if the slider is meant to be `Tracking` and it's not...
- call the sliderend's `MissForcefully()` method

If there's a better way of doing this (or if the above just sounds incredibly janky) then I'm eager to hear alternatives

**Comment by smoogipoo** [comment]
Sounds like a better alternative to https://github.com/ppy/osu/discussions/16552

**Reply by TheFantasticLoki** [reply]
I really REALLY like this idea as a mod especially as another variant of hard rock similar to how nightcore functions. Always felt there was a lack of difficulty increase mods.

**Comment by smoogipoo** [comment]
I think your path of implementation for this is pretty spot on with what needs to be done, give it a shot :)

**Comment by apollo-dw** [comment]
Hey hey, here's a WIP version of this mod: https://github.com/apollo-dw/osu/tree/strict-tracking

Currently, this is functional (and it was very fun to figure my way around the code), however as per my consultations with bdach, this implementation isn't fit for review. Reasons why:

1. Mods aren't really supposed to require modifying object classes, which I have totally done here
2. I'd have to make `DisplayResult` settable to display a miss when the slider stops being tracked

My understanding is that the mod would require more flexibility than is currently available? Either way, the branch is up above for play testing and getting a feel, just be wary of sliderbreaks looking weird because of point 2 above wink wonk

**Reply by smoogipoo** [reply]
It's kinda hard not to modify object classes when you're affecting judgements, classic mod does the same thing and I have no better suggestion for the time being. Though I wonder if there's a reason you want the tail to only be judged as a tick? Is it just to make it more lenient?

**Reply by apollo-dw** [reply]
Well without doing that, sliderbreaking with the mod doesn't drop combo, as `MissForcefully` feeds a `MinResult` judgement.

**Reply by bdach** [reply]
Just to clarify the above, slider tails use the "small ticks" family of hit results which indeed do not drop combo. Another issue is that they do not have a visible judgement which is also a bit of a problem.

If mods were able to inject their own hitobject types then this would be resolveable rather amicably, but that flexibility isn't there right now.

**Reply by smoogipoo** [reply]
Aha, makes sense.

> If mods were able to inject their own hitobject types

Now that I think about it, is there an issue preventing this that isn't visible at face value? Could be done by implementing `IApplicableAfterBeatmapConversion` and replacing the hitobjects directly in the given beatmap. You'd just have to copy over all of these properties:
![image](https://user-images.githubusercontent.com/1329837/157582065-fb31002b-a912-46ce-8b66-016db768a1a9.png)

Don't even need to call `ApplyDefaults()` since that's done automatically later on, so it should be pretty speedy too.

**Reply by bdach** [reply]
> is there an issue preventing this that isn't visible at face value?

Pooling would be one. There is no way to register a pool of drawable hitobjects from a mod.

**Reply by peppy** [reply]
Would registering a pool be required if they are using an existing drawable representation though?

**Reply by smoogipoo** [reply]
Yeah as @peppy said. It seems like the pooling mechanism can use the base-typed pools for derived `HitObject`s:
https://github.com/ppy/osu/blob/c17dba4948ac7ed76a6d08b1a93a0728ab82b331/osu.Game/Rulesets/UI/Playfield.cs#L337-L348

If this works, then I would love to see classic mod do the same, which brings up a bit of a different problem - that the two mods would become incompatible as they would add their own slider types. Which is why I'm not sure whether this is the best path as opposed to modifying the classes themselves as classic mod already does.

**Reply by bdach** [reply]
I forgot that that logic was there - but this mod would want its own drawable object implementation too ideally, to show judgements on slider tails.

As for the incompatibility with classic, that will have to happen either way, because one of the options provided by classic is to enable judging the slider as a whole proportionally to number of nested objects hit, which would break this mod.

I don't know, I for one am not looking forward to a future where mod implementors are free to be placing boolean switches everywhere in objects and drawable objects. Classic is a special exemption to me, because of how integral it is to onboarding stable users, but I wouldn't like to see any more of that.

**Reply by smoogipoo** [reply]
> As for the incompatibility with classic, that will have to happen either way

I would hope that at least some of classic mod is compatible with this. For example players wanting to turn off slider head accuracy but maintain strict tracking and the new notelock. I don't see it as a huge issue if there's two judgements in this case - it means that maybe sliders become a little bit harser (by 30 score) but both mods become fully compatible and no changes should be required to either mod.

**Reply by bdach** [reply]
Proportional judgement of sliders would break the primary objective of this mod (that leaving the slider ball tracking area should result in a miss and dropped combo). You'd probably get 50s or 100s instead. There's no way to bypass that unless you permit this mod to override classic behaviour which is not an avenue we want to go down I don't think.

**Reply by smoogipoo** [reply]
It will still result in a miss and dropped combo as a result of missing the "tick" (the tail). But it will also result in a 100 for the slider's own judgement (I believe), which I have no issue with.

**Reply by bdach** [reply]
I guess if you're really fine with more boolean flags in hitobjects then I'll yield on this. But it has long been a rule that new mods should not be changing logic in hitobject classes unless it is absolutely unavoidable. I strongly agree with that rule, and think that adding the boolean flags that this mod would need would not be tenable long term.

**Reply by apollo-dw** [reply]
Hi, I'm just re-poking to see what the intended direction for implementation would look like now. Is `JudgeAsSliderTick` correct, in terms of as a boolean flag in `SliderTailCircle` and the decision to use slidertick judgement? Note the reprecussions on the results screen, for example. Also still a little unsure about the drawable judgement display side of things.

**Reply by smoogipoo** [reply]
The answer to that question depends on what you want. Do you want the mod to be compatible with the classic mod? If not then to reduce friction I'd use a derived class as per above suggestions, because there are some very strong voices calling for mods to never set hitobject properties.
But if you do want it to be compatible, then we need to figure out how to actually do this.

As for the drawable part I again need opinions from one of those proposing mods to never set any properties to explain how they'd do it, because I'm all out of ideas. @bdach

**Reply by bdach** [reply]
I thought I've explained my view on this exhaustively before - custom object inheriting slider with custom drawable hitobject inheriting drawable slider would give an opportunity to override the `DisplayResult` flag on the DHO.

I could agree with exposing `DrawableHitObject.DisplayResult` as settable by mods because that's actually something that could be used by more mods and it's at a high enough level wherein it's not a huge problem to me. But that still doesn't solve the other half of the problem (changing slider judgements) and I think doing that via a boolean switch specific to this one mod would be a bad precedent (again, classic is a special exception to me).

I don't know. Would be interested in thoughts from @peppy too as he's previously expressed the same view that mods ideally should not expose new settable properties on hitobjects and be as self contained as feasible.
